doc/developer-guide/documentation/RPC.en.rst

```diff
@@ -0,0 +1,397 @@
+
+Sequence diagram for a command sent from :program:`traffic_ctl` to when it is recieved by a plugin. 
+
+.. image:: ../../uml/images/RPC-sequence-diagram.svg
+    
+.. note::
+
+    Currently a fire and forget model. traffic_manager sends out an asynchoronous signal without any acknowledgement. It then proceeds to send a response itself. 
+
+=======================
+Remote RPC vs Local RPC 
+=======================
+
+The RPC API for remote clients, such as :program:`traffic_ctl`, :program:`traffic_cop`, etc, is different from the RPC API used between :program:`traffic_manager` and :program:`traffic_server`. 
+
+:program:`traffic_manager` acts like a bridge for remote clients to interact with :program:`traffic_server`. Thus, it is currently impossible to do things like have :program:`traffic_ctl` directly send messages to :program:`traffic_server`. Classes suffixed with "Remote", ie. :file:`CoreAPIRemote.cc`, and classes suffixed with "Local", ie. :file:`NetworkUtilsLocal.cc` are for remote and local clients, respectively. The following sections will note which set of RPC's are relevant.
+
+=======================
+Serialization Mechanism
+=======================
+
+.. class:: MgmtMarshall
+
+   This is the class used to marshall data objects. It provides functions to marshall and unmarshall data. Each data object is associated with a field. Fields are ``MgmtMarshallType``:
+
+.. type:: MgmtMarshallType
+
+.. c:macro:: MGMT_MARSHALL_INT
+.. c:macro:: MGMT_MARSHALL_LONG
+.. c:macro:: MGMT_MARSHALL_STRING
+.. c:macro:: MGMT_MARSHALL_DATA
+
+Serialized data is sent over sockets. This is the general serialization mechanism for RPC communcation. 
+
+Generally, for remote clients sending messages to :program:`traffic_server`, the message is serialized using remote RPC APIs. The serialized message is send to :program:`traffic_manager` and :program:`traffic_manager` then simply relays the message to :program:`traffic_server`. :program:`traffic_server` eventually unserializes the message. 
+
+Details for how :program:`traffic_manager` and :program:`traffic_server` communicate are documented in the next section.
+
+Marshalling:
+============
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_marshall(void *buf, size_t remain, const MgmtMarshallType *fields, unsigned count, ...)
+
+   Variable argument wrapper for ``mgmt_message_marshall_v``. Allows for different datatypes to be marshalled together as long as a field is specified for each data object. Arguments should be references to objects to be marshalled. 
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_marshall_v(void *buf, size_t remain, const MgmtMarshallType *fields, unsigned count, va_list ap)
+
+   This function goes through all the data objects and serializes them together into a buffer. Based on the field, the number of bytes is determined and if there is enough space, it is written into the buffer. 
+
+    * MGMT_MARSHALL_INT: 4 bytes.
```

Comment:
These should have the same indent level as the preceeding paragraph.

Reply:
updated.